File: 4.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int m = nums1.length;
        int n = nums2.length;
        if (m > n) return findMedianSortedArrays(nums2, nums1);
        if (m == 0) return findMedian(nums2);
        if (n == 0) return findMedian(nums1);
        int n1 = 0;
        int start = 0, end = m;
        int n2 = (m+n+1)/2-n1;
        double res1 = 0;
        double res2 = 0;
        while (start <= end) {
            n1 = start+(end-start)/2;
            n2 = (m+n+1)/2-n1;
            int l1 = n1 == 0 ? Integer.MIN_VALUE : nums1[n1-1];
            int r1 = n1 == m ? Integer.MAX_VALUE : nums1[n1];
            int l2 = n2 == 0 ? Integer.MIN_VALUE : nums2[n2-1];
            int r2 = n2 == n ? Integer.MAX_VALUE : nums2[n2];
            if (l1 > r2) end = n1-1;
            else if (l2 > r1) start = n1+1;
            else {
                res1 = Math.max(l1, l2);
                res2 = Math.min(r1, r2);
                break;
            }
        }
        return (m+n) % 2 == 0 ? res1/2+res2/2 : res1;
    }

    private double findMedian(int[] nums) {
        int n = nums.length;
        if (n == 0) return 0;
        if (n % 2 == 0) return (double) (nums[n/2]+nums[n/2-1])/2;
        return (double) nums[n/2];
    }
}

File: 4.py
class Solution:
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        if len(nums1) < len(nums2):
            nums1, nums2 = nums2, nums1
        len1 = len(nums1)
        len2 = len(nums2)
        lower = 0
        upper = 2*len2
        while lower <= upper:
            j = int((lower+upper)/2)
            i = int(len1+len2-j)
            if i == 0:
                L1 = float("-inf")
            else:
                L1 = nums1[int((i-1)/2)]
            R1 = nums1[int(i/2)] if i != 2*len1 else float("inf")
            L2 = nums2[int((j-1)/2)] if j != 0 else float("-inf")
            R2 = nums2[int(j/2)] if j != 2*len2 else float("inf")
            if L1 > R2:
                lower += 1
            elif R1 < L2:
                upper -= 1
            else:
                return (max(L1,L2)+min(R1,R2))/2 